File: frontend/src/services/quota.ts
import api from './api';

// 配额信息接口
export interface QuotaInfo {
  user_id: number;
  daily_quota: number;
  daily_used: number;
  daily_remaining: number;
  monthly_quota: number;
  monthly_used: number;
  monthly_remaining: number;
  permanent_quota: number;
  total_available: number;
  is_vip: boolean;
  vip_expire_at?: string;
  daily_reset_at: string;
  monthly_reset_at: string;
  total_consumed: number;
  last_consume_at?: string;
  last_reset_at: string;
  next_reset_at: string;
}

// 配额日志接口
export interface QuotaLog {
  id: number;
  user_id: number;
  quota_type: 'daily' | 'monthly' | 'permanent';
  amount: number;
  service_type: 'chat' | 'generate' | 'grade' | 'diagnose' | 'ocr';
  description?: string;
  created_at: string;
}

// 配额使用记录（别名，兼容性）
export interface QuotaUsageRecord extends QuotaLog {}

// 充值记录接口
export interface RechargeLog {
  id: number;
  user_id: number;
  quota_type: 'daily' | 'monthly' | 'permanent';
  amount: number;
  reason?: string;
  operator_id?: number;
  recharge_method: 'manual' | 'purchase' | 'reward' | 'vip';
  order_id?: string;
  created_at: string;
}

// 配额服务类
class QuotaService {
  /**
   * 获取我的配额信息
   */
  async getMyQuota(): Promise<QuotaInfo> {
    const response = await api.get('/api/v1/quota/my');
    return response.data;
  }

  /**
   * 获取配额使用日志
   */
  async getQuotaLogs(page: number = 1, pageSize: number = 20): Promise<{
    list: QuotaLog[];
    total: number;
    page: number;
    size: number;
  }> {
    const response = await api.get('/api/v1/quota/logs', {
      params: { page, page_size: pageSize }
    });
    return response.data;
  }

  /**
   * 获取配额使用历史（别名，兼容性）
   */
  async getUsageHistory(page: number = 1, pageSize: number = 20): Promise<{
    records: QuotaUsageRecord[];
    total: number;
    page: number;
    page_size: number;
  }> {
    const result = await this.getQuotaLogs(page, pageSize);
    return {
      records: result.list,
      total: result.total,
      page: result.page,
      page_size: result.size,
    };
  }

  /**
   * 获取充值记录
   */
  async getRechargeLogs(page: number = 1, pageSize: number = 20): Promise<{
    list: RechargeLog[];
    total: number;
    page: number;
    size: number;
  }> {
    const response = await api.get('/api/v1/quota/recharge-logs', {
      params: { page, page_size: pageSize }
    });
    return response.data;
  }

  /**
   * 检查配额是否足够
   */
  async checkQuota(amount: number): Promise<{
    sufficient: boolean;
    amount: number;
  }> {
    const response = await api.get('/api/v1/quota/check', {
      params: { amount }
    });
    return response.data;
  }

  /**
   * 充值配额（管理员）
   */
  async rechargeQuota(data: {
    user_id: number;
    quota_type: 'daily' | 'monthly' | 'permanent';
    amount: number;
    reason?: string;
    method?: 'manual' | 'purchase' | 'reward' | 'vip';
  }): Promise<void> {
    await api.post('/api/v1/quota/recharge', data);
  }

  /**
   * 设置VIP（管理员）
   */
  async setVIP(data: {
    user_id: number;
    days: number;
    extra_quota: number;
  }): Promise<void> {
    await api.post('/api/v1/quota/vip', data);
  }

  /**
   * 取消VIP（管理员）
   */
  async cancelVIP(userId: number): Promise<void> {
    await api.delete(`/api/v1/quota/vip/${userId}`);
  }

  /**
   * 格式化配额类型
   */
  formatQuotaType(type: string): string {
    const map: { [key: string]: string } = {
      'daily': '日配额',
      'monthly': '月配额',
      'permanent': '永久配额'
    };
    return map[type] || type;
  }

  /**
   * 格式化服务类型
   */
  formatServiceType(type: string): string {
    const map: { [key: string]: string } = {
      'chat': 'AI对话',
      'generate': '生成题目',
      'grade': '批改答案',
      'diagnose': '学习诊断',
      'ocr': 'OCR识别'
    };
    return map[type] || type;
  }

  /**
   * 格式化充值方式
   */
  formatRechargeMethod(method: string): string {
    const map: { [key: string]: string } = {
      'manual': '手动充值',
      'purchase': '购买充值',
      'reward': '奖励',
      'vip': 'VIP赠送'
    };
    return map[method] || method;
  }

  /**
   * 获取配额不足提示
   */
  getInsufficientQuotaMessage(quotaInfo: QuotaInfo): string {
    if (quotaInfo.permanent_quota > 0) {
      return `永久配额剩余: ${quotaInfo.permanent_quota}`;
    }

    if (quotaInfo.daily_remaining > 0) {
      return `今日剩余配额: ${quotaInfo.daily_remaining}/${quotaInfo.daily_quota}`;
    }

    if (quotaInfo.monthly_remaining > 0) {
      return `本月剩余配额: ${quotaInfo.monthly_remaining}/${quotaInfo.monthly_quota}`;
    }

    return '配额已用尽，请充值或等待重置';
  }

  /**
   * 计算下次重置时间
   */
  getNextResetTime(quotaInfo: QuotaInfo): string {
    const now = new Date();
    const nextReset = new Date(quotaInfo.next_reset_at || quotaInfo.daily_reset_at);
    const diff = nextReset.getTime() - now.getTime();

    if (diff < 0) {
      return '即将重置';
    }

    const hours = Math.floor(diff / (1000 * 60 * 60));
    const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));

    if (hours > 24) {
      const days = Math.floor(hours / 24);
      return `${days}天后重置`;
    }

    if (hours > 0) {
      return `${hours}小时${minutes}分钟后重置`;
    }

    return `${minutes}分钟后重置`;
  }

  /**
   * 格式化配额显示
   */
  formatQuota(quota: QuotaInfo): string {
    return `日配额: ${quota.daily_remaining}/${quota.daily_quota} | 月配额: ${quota.monthly_remaining}/${quota.monthly_quota}${
      quota.permanent_quota > 0 ? ` | 永久: ${quota.permanent_quota}` : ''
    }`;
  }

  /**
   * 获取配额状态
   */
  getQuotaStatus(quota: QuotaInfo): 'sufficient' | 'low' | 'depleted' {
    if (quota.total_available === 0) {
      return 'depleted';
    }
    if (quota.total_available <= 10) {
      return 'low';
    }
    return 'sufficient';
  }
}

export default new QuotaService();

File: backend/internal/domain/models/quota.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// QuotaType 配额类型
type QuotaType string

const (
	QuotaTypeDaily     QuotaType = "daily"     // 日配额
	QuotaTypeMonthly   QuotaType = "monthly"   // 月配额
	QuotaTypePermanent QuotaType = "permanent" // 永久配额
)

// ServiceType AI服务类型
type ServiceType string

const (
	ServiceTypeChat     ServiceType = "chat"     // AI对话
	ServiceTypeGenerate ServiceType = "generate" // 生成题目
	ServiceTypeGrade    ServiceType = "grade"    // 批改答案
	ServiceTypeDiagnose ServiceType = "diagnose" // 学习诊断
	ServiceTypeOCR      ServiceType = "ocr"      // OCR识别
)

// RechargeMethod 充值方式
type RechargeMethod string

const (
	RechargeMethodManual   RechargeMethod = "manual"   // 手动充值
	RechargeMethodPurchase RechargeMethod = "purchase" // 购买充值
	RechargeMethodReward   RechargeMethod = "reward"   // 奖励
	RechargeMethodVIP      RechargeMethod = "vip"      // VIP赠送
)

// UserAIQuota 用户AI配额模型
type UserAIQuota struct {
	ID     uint `gorm:"primarykey" json:"id"`
	UserID uint `gorm:"not null;uniqueIndex" json:"user_id"`

	// 日配额
	DailyQuota   int       `gorm:"not null;default:10;check:daily_quota >= 0" json:"daily_quota"`
	DailyUsed    int       `gorm:"not null;default:0;check:daily_used >= 0" json:"daily_used"`
	DailyResetAt time.Time `gorm:"not null" json:"daily_reset_at"`

	// 月配额
	MonthlyQuota   int       `gorm:"not null;default:300;check:monthly_quota >= 0" json:"monthly_quota"`
	MonthlyUsed    int       `gorm:"not null;default:0;check:monthly_used >= 0" json:"monthly_used"`
	MonthlyResetAt time.Time `gorm:"not null" json:"monthly_reset_at"`

	// 永久配额（不重置）
	PermanentQuota int `gorm:"not null;default:0;check:permanent_quota >= 0" json:"permanent_quota"`

	// VIP相关
	IsVIP        bool       `gorm:"default:false" json:"is_vip"`
	VIPExpireAt  *time.Time `json:"vip_expire_at,omitempty"`

	// 统计字段
	TotalConsumed  int64      `gorm:"not null;default:0" json:"total_consumed"`
	LastConsumeAt  *time.Time `json:"last_consume_at,omitempty"`

	// 审计字段
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// TableName 指定表名
func (UserAIQuota) TableName() string {
	return "user_ai_quotas"
}

// GetAvailableQuota 获取可用配额
func (q *UserAIQuota) GetAvailableQuota() int {
	dailyAvailable := q.DailyQuota - q.DailyUsed
	monthlyAvailable := q.MonthlyQuota - q.MonthlyUsed
	return q.PermanentQuota + monthlyAvailable + dailyAvailable
}

// GetDailyRemaining 获取今日剩余配额
func (q *UserAIQuota) GetDailyRemaining() int {
	remaining := q.DailyQuota - q.DailyUsed
	if remaining < 0 {
		return 0
	}
	return remaining
}

// GetMonthlyRemaining 获取本月剩余配额
func (q *UserAIQuota) GetMonthlyRemaining() int {
	remaining := q.MonthlyQuota - q.MonthlyUsed
	if remaining < 0 {
		return 0
	}
	return remaining
}

// NeedsDailyReset 检查是否需要重置日配额
func (q *UserAIQuota) NeedsDailyReset() bool {
	return time.Now().After(q.DailyResetAt)
}

// NeedsMonthlyReset 检查是否需要重置月配额
func (q *UserAIQuota) NeedsMonthlyReset() bool {
	return time.Now().After(q.MonthlyResetAt)
}

// AIQuotaLog AI配额使用日志
type AIQuotaLog struct {
	ID     uint `gorm:"primarykey" json:"id"`
	UserID uint `gorm:"not null;index:idx_ai_quota_logs_user_created" json:"user_id"`

	// 配额类型
	QuotaType QuotaType `gorm:"type:varchar(20);not null;index" json:"quota_type"`

	// 消耗数量
	Amount int `gorm:"not null;check:amount > 0" json:"amount"`

	// 服务类型
	ServiceType ServiceType `gorm:"type:varchar(50);not null;index" json:"service_type"`

	// 描述信息
	Description string `gorm:"type:text" json:"description,omitempty"`

	// 请求元数据
	RequestID string `gorm:"type:varchar(100)" json:"request_id,omitempty"`
	IPAddress string `gorm:"type:inet" json:"ip_address,omitempty"`
	UserAgent string `gorm:"type:text" json:"user_agent,omitempty"`

	// 审计字段
	CreatedAt time.Time      `gorm:"index:idx_ai_quota_logs_user_created" json:"created_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// TableName 指定表名
func (AIQuotaLog) TableName() string {
	return "ai_quota_logs"
}

// QuotaRechargeLog 配额充值记录
type QuotaRechargeLog struct {
	ID     uint `gorm:"primarykey" json:"id"`
	UserID uint `gorm:"not null;index" json:"user_id"`

	// 充值配额类型
	QuotaType QuotaType `gorm:"type:varchar(20);not null" json:"quota_type"`

	// 充值数量
	Amount int `gorm:"not null;check:amount > 0" json:"amount"`

	// 充值原因
	Reason string `gorm:"type:varchar(100)" json:"reason,omitempty"`

	// 操作者ID
	OperatorID *uint `json:"operator_id,omitempty"`

	// 充值方式
	RechargeMethod RechargeMethod `gorm:"type:varchar(20);not null;default:'manual'" json:"recharge_method"`

	// 订单号
	OrderID string `gorm:"type:varchar(100);index" json:"order_id,omitempty"`

	// 审计字段
	CreatedAt time.Time      `gorm:"index" json:"created_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	User     *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Operator *User `gorm:"foreignKey:OperatorID" json:"operator,omitempty"`
}

// TableName 指定表名
func (QuotaRechargeLog) TableName() string {
	return "quota_recharge_logs"
}

// QuotaInfo 配额信息（用于API响应）
type QuotaInfo struct {
	// 配额统计
	DailyQuota     int `json:"daily_quota"`
	DailyUsed      int `json:"daily_used"`
	DailyRemaining int `json:"daily_remaining"`

	MonthlyQuota     int `json:"monthly_quota"`
	MonthlyUsed      int `json:"monthly_used"`
	MonthlyRemaining int `json:"monthly_remaining"`

	PermanentQuota int `json:"permanent_quota"`
	TotalAvailable int `json:"total_available"`

	// VIP信息
	IsVIP       bool       `json:"is_vip"`
	VIPExpireAt *time.Time `json:"vip_expire_at,omitempty"`

	// 重置时间
	DailyResetAt   time.Time `json:"daily_reset_at"`
	MonthlyResetAt time.Time `json:"monthly_reset_at"`

	// 统计
	TotalConsumed int64      `json:"total_consumed"`
	LastConsumeAt *time.Time `json:"last_consume_at,omitempty"`
}

// ToQuotaInfo 转换为QuotaInfo
func (q *UserAIQuota) ToQuotaInfo() *QuotaInfo {
	return &QuotaInfo{
		DailyQuota:       q.DailyQuota,
		DailyUsed:        q.DailyUsed,
		DailyRemaining:   q.GetDailyRemaining(),
		MonthlyQuota:     q.MonthlyQuota,
		MonthlyUsed:      q.MonthlyUsed,
		MonthlyRemaining: q.GetMonthlyRemaining(),
		PermanentQuota:   q.PermanentQuota,
		TotalAvailable:   q.GetAvailableQuota(),
		IsVIP:            q.IsVIP,
		VIPExpireAt:      q.VIPExpireAt,
		DailyResetAt:     q.DailyResetAt,
		MonthlyResetAt:   q.MonthlyResetAt,
		TotalConsumed:    q.TotalConsumed,
		LastConsumeAt:    q.LastConsumeAt,
	}
}
